Add block-spin update sampler

kevinchern opened the pull request:
This PR introduces a block-spin update sampler.
It includes duplicate code from https://github.com/dwavesystems/dwave-pytorch-plugin/pull/40

jackraymond commented on dwave/plugins/torch/samplers/block_spin_update.py:
```diff
@@ -0,0 +1,165 @@
+    @torch.compile
+    @torch.no_grad
+    def step_(self, beta: torch.Tensor):
+        """Performs a block-Gibbs update in-place.
```
or block-Metropolis

jackraymond commented on dwave/plugins/torch/samplers/block_spin_update.py:
```diff
@@ -0,0 +1,165 @@
+        grbm (GRBM): The Graph-Restricted Boltzmann Machine to sample from.
+    """
+
+    def __init__(self, G: Graph, grbm: GRBM, num_reads: int, kind: str):
```
I think 'kind' is called 'proposal_acceptance_criteria' in SimulatedAnnealingSampler. It took me a while to work out what kind was, and would be good to standardize.

jackraymond replied:
Add 'Gibbs' as default

jackraymond commented on dwave/plugins/torch/samplers/block_spin_update.py:
```diff
@@ -0,0 +1,165 @@
+            lin2lattice = dnx.pegasus_coordinates(self.G.graph['rows']).linear_to_pegasus
+            crayon = dnx.pegasus_four_color
+        elif topology == "chimera":
+            lin2lattice = dnx.chimera_coordinates(self.G.graph['rows']).linear_to_chimera
```
chimera graphs support 2 node types, this is not required if the graph is already coordinated.

kevinchern commented on dwave/plugins/torch/samplers/block_spin_update.py:
```diff
@@ -0,0 +1,165 @@
+            h = self.linear[block]
+            J = self.quadratic[self.padded_adjacencies_weight[block]]
+            effective_field = (xnbr * J.unsqueeze(0)).sum(2) + h
+            if self.kind == "metropolis":
```
Changed to title case and made input case-insensitive, i.e., `self.kind = kind.title()`

VolodyaCO commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,227 @@
+
+    Args:
+        grbm (GRBM): The Graph-Restricted Boltzmann Machine to sample from.
+        crayon (Callable): A colouring function; a function that maps a single node of the `grbm` to
```
Callable[[int], int], to better specify the function signature.

kevinchern replied:
Using Hashable instead

VolodyaCO commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,227 @@
+    def __init__(self, grbm: GRBM, crayon: Callable, num_chains: int,
+                 proposal_acceptance_criteria: Literal["Gibbs", "Metropolis"] = "Gibbs", seed=None):
+        super().__init__()
+        if num_chains < 1 or not isinstance(num_chains, int):
```
python performs these checks in order from left to right. Therefore, if you pass `num_chains="hello world"`, a different error will be thrown saying that you cannot compare order between a str with an int. First check type, then check value.

VolodyaCO commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,227 @@
+
+    def _get_adjacencies(self) -> tuple[list[torch.Tensor], list[torch.Tensor]]:
+        """Create two lists of padded adjacency lists (tensors), one for neighbouring indices and
+        another for the corresponding weight indices.
```
I got a bit confused when reading this method. "edge indices" seems more intuitive for me than "weight indices", according to what is done in this method.

kevinchern replied:
It hasn't been that long since I wrote the code and I'm already confused by the documentation as well. Updated documentation.

kevinchern commented on dwave/plugins/torch/tensor.py:
```diff
@@ -0,0 +1,38 @@
+__all__ = ["rands"]
+
+
+def rands(size: _size, **kwargs) -> torch.Tensor:
```
> I would recommend using a more descriptive name such as random_spins

How does `randspin` sound? (Following torch conventions, e.g., `randn` for normal, `rand` for uniform, and `randint` for integers

kevinchern commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,227 @@
+    """A block-spin update sampler for graph-restricted Boltzmann machines.
+
+    Note this is a sampler and not a neural network. It extends `nn.Module` for the convenience of
+    managing devices of indices (stored as parameters).
```
Good point! Thanks for this---the `to` method will be addressed in the PR for https://github.com/dwavesystems/dwave-pytorch-plugin/issues/52

kevinchern commented:
Addressed @thisac and @VolodyaCO 's PRs. Briefly,

1. Remove inheritance of `nn.Module`
2.  Add functional module docstring
3. Add explicit named args for, and renamed to `randspin`
4. Improved documentation and type-hinting

jackraymond commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,246 @@
+    efficiently sample in blocks of spins. Ideally, an adjacency list can be used, however,
+    adjacencies are ragged, making vectorization inapplicable.
+
+    Block-Gibbs and Block-Metropolis obey detailed balance and are ergodic methods at finite
```
Block-Metropolis is not ergodic at zero temperature. Change to 'finite non-zero'

jackraymond commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,246 @@
+    Block-Gibbs and Block-Metropolis obey detailed balance and are ergodic methods at finite
+    temperature which, at fixed parameters, converge upon Boltzmann distributions. Block-Metropolis
+    allows higher acceptance rates for proposals (faster single-step mixing), but is non-ergodic in
+    the limit of zero temperature. Decorrelation from an initial condition can be slower.
```
can be non-ergodic in the limit of zero or infinite temperature.

kevinchern commented:
> Looks good tests ran for me.
> 
> I'd recommend you allow setting of x as an initial condition. Check it doesn't move when num_sweeps is none, and check it oscillates under BlockMetropolis when temperature is infinity as tests perhaps.

In the latest commit, I made the following changes:

- Fixed documentation RE nonzero & finite temperature regimes.
- Allow setting of initial states.
    - Test by checking `sampler.x.tolist()` is equal to input initial states list
- Added custom errors to distinguish from `ValueError`

@jackraymond @thisac

thisac commented on dwave/plugins/torch/nn/functional.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def bit2spin_soft(b: torch.Tensor) -> torch.Tensor:
+    """Maps input `b` to `2b-1`.
```
Double back-ticks for code.
```suggestion
    """Maps input ``b`` to ``2b-1``.
```
Same for rest of docstring and file.

kevinchern replied:
Changed to (and other occurrences)
```
:math:`b`
```

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+            "Gibbs" or "Metropolis".
+    """
+
+    def to(self, device: DeviceLikeType) -> None:
```
I feel like this should either work in the same was torch's `to()` works (i.e., returns a copy of self on the new device) _or_ have a different name to do distinguish it (like `to_inplace()`).

kevinchern replied:
Added a `return self` statement. `nn.Module.to` move weights in-place.
```
from torch.nn import Linear
A = Linear(3, 2)
print(A.weight.device)
print(A.to("meta").weight.device)
print(A.weight.device)
```
Confusingly, `torch.Tensor` and `nn.Parameter` do not get moved in-place.

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+        if not isinstance(num_chains, int) or num_chains < 1:
+            raise NonPositiveIntegerError(
+                "Number of reads should be a positive integer."
+            )
```
Since this is both a `TypeError` and `ValueError` check you could have `NonPositiveIntegerError` inherit from both, which would allow it to be caught by both.
```python
class NonPositiveIntegerError(TyperError, ValueError):
    """TODO"""
```

kevinchern replied:
(Removed per other comment)

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+        self._crayon = crayon
+        if not self._valid_crayon():
+            raise InvalidColouringError(
+                "``crayon`` is not a valid colouring of ``grbm``. "
```
No need for double back-ticks here. I usually just use `'` since they look nice and clear.

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+        self._rng = None
+        if seed is not None:
+            self._rng = torch.Generator()
+            self._rng.manual_seed(seed)
```
(minor) IMO slightly clearer.
```suggestion
        if seed is not None:
            self._rng = torch.Generator()
            self._rng.manual_seed(seed)
        else:
            self._rng = None
```

kevinchern replied:
I think the original is cleaner 😝

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+        return initial_states
+
+    def _valid_crayon(self) -> bool:
+        """Determines whether `crayon` is a valid colouring of `grbm`.
```
```suggestion
        """Determines whether ``crayon`` is a valid colouring of ``grbm``.
```

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+        return True
+
+    def _get_partition(self) -> nn.ParameterList:
+        """Computes the vertex partition induced by the colouring function `crayon`.
```
```suggestion
        """Computes the vertex partition induced by the colouring function ``crayon``.
```

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+        self._x[:, block] = self._x[:, block]*flip
+
+    @torch.no_grad
+    def _gibbs_update(self, beta, block, effective_field):
```
Missing type hints.

thisac commented on dwave/plugins/torch/tensor.py:
```diff
@@ -0,0 +1,64 @@
+
+    Raises:
+        ValueError: If `low` is supplied as a keyword argument.
+        ValueError: If `high` is supplied as a keyword argument.
```
```suggestion
        ValueError: If ``low`` is supplied as a keyword argument.
        ValueError: If ``high`` is supplied as a keyword argument.
```

thisac commented on dwave/plugins/torch/tensor.py:
```diff
@@ -0,0 +1,64 @@
+                           dtype=dtype,
+                           device=device,
+                           requires_grad=requires_grad,
+                           **kwargs)-1
```
Some cleanup needed. Spaces around `-` and `*`, and perhaps separating out the `torch.randint()` call since the final `2*` and `-1` get kind-of lost in there.

kevinchern replied:
> `-1` get lost

Changed to `b = randint ... ` and `return 2 * b - 1`

kevinchern replied:
> Also not convinced that this i needed. Couldn't you just use the regular torch.randint() and pass the results through bit2spin_soft()? Other than that this only sets low/high which isn't too cumbersome to do manually.

My motivation for adding it is similar to adding a `bit2spin` function. It's very simple to write repeatedly, but it helps with readability to have a dedicated function. What are some reasons to avoid adding this?

kevinchern commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+
+
+class InvalidColouringError(ValueError):
+    pass
```
Removed all custom errors in favour of `assertRaisesRegex` as these errors were used only in testing

kevinchern commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,348 @@
+
+        If the device is "meta", then the random number generator ``BlockSpinSampler._rng`` (RNG)
+        will not be modified at all. For all other devices, all fields ``_x``, ``_zeros``,
+        ``_schedule``, ``_partition``, ``_padded_adjacencies``, and ``_padded_adjacencies_weight``
```
Replaced references to hidden attributes with descriptive names.

kevinchern commented on tests/test_block_spin_sampler.py:
```diff
@@ -0,0 +1,279 @@
+
+    GRBM_CRAYON_TEST_CASES = [(GRBM_ZEPHYR, CRAYON_ZEPHYR),
+                              (GRBM_BIPARTITE, CRAYON_BIPARTITE),
+                              (GRBM_SINGLE, CRAYON_SINGLE)]
```
If I understood correctly, that would be incompatible with `@parameterized.expand`

jackraymond reviewed (approved):
No changes necessary 

Test ran fine for me, got it working for my benchmarking straightfowardly.

So far as I can tell on an L4:
156 ns per spin update for single sample on Zephyr[m=12] problem.
drops to ~ 156/58 ns per spin update with 58 fold parallelism on an L4. Which is good.

thisac commented on dwave/plugins/torch/samplers/block_spin_sampler.py:
```diff
@@ -0,0 +1,330 @@
+            x (torch.Tensor): A tensor of shape (batch_size, dim) interpreted as a batch of
+                partially-observed spins. Entries marked with ``torch.nan`` will be sampled; entries
+                with +/-1 values will remain constant.
+        Returns:
```
```suggestion
                with +/-1 values will remain constant.

        Returns:
```
